src/librustc_jobserver/lib.rs

```diff
@@ -0,0 +1,292 @@
+}
+
+const EMPTY: usize = 0;
+const CARGO_REQUESTED: usize = 1;
+const MAKE_REQUESTED: usize = 2;
+static TOKEN_REQUESTS: AtomicUsize = AtomicUsize::new(0);
+
+fn should_notify() -> bool {
+    let value = TOKEN_REQUESTS.load(Ordering::SeqCst);
+    assert!(value != EMPTY, "jobserver must be initialized");
+    value == CARGO_REQUESTED
+}
+
+/// This will only adjust the global value to the new value of token_requests
+/// the first time it's called, which means that you want to make sure that the
+/// first call you make has the right value for `token_requests`. We try to help
+/// out a bit by making sure that this is called before any interaction with the
+/// jobserver (which usually happens almost immediately as soon as rustc does
+/// anything due to spawning up the Rayon threadpool).
+///
+/// Unfortunately the jobserver is inherently a global resource (we can't
+/// have more than one) so the token requesting strategy must likewise be global.
+///
+/// Usually this doesn't matter too much, as you're not wanting to set the token
+/// requests unless you're in the one-rustc-per-process model, and we help out
+/// here a bit by not resetting it once it's set (i.e., only the first init will
+/// change the value).
+pub fn initialize(token_requests: bool) {
+    lazy_static::initialize(&GLOBAL_CLIENT);
+    lazy_static::initialize(&HELPER);
+    let previous = TOKEN_REQUESTS.compare_and_swap(
+        EMPTY,
+        if token_requests { CARGO_REQUESTED } else { MAKE_REQUESTED },
+        Ordering::SeqCst,
+    );
+    if previous == EMPTY {
+        log::info!("initialized rustc jobserver, set token_requests={:?}", token_requests);
+    }
+}
+
+struct TokenCounts {
+    used: usize,
+    pending: usize,
+}
+
+pub struct Helper {
+    tokens: Arc<Mutex<TokenCounts>>,
+    waiter: Arc<Condvar>,
+    helper: Mutex<jobserver_crate::HelperThread>,
+}
+
+impl Helper {
+    fn new() -> Self {
+        let tokens = Arc::new(Mutex::new(TokenCounts { used: 1, pending: 0 }));
+        let waiter = Arc::new(Condvar::new());
+        let tokens_thread = tokens.clone();
+        let waiter_thread = waiter.clone();
+        let helper = GLOBAL_CLIENT
+            .clone()
+            .into_helper_thread(move |token| {
+                log::trace!("Helper thread token sending into channel");
+                // We've acquired a token, but we need to not use it as we have our own
+                // custom release-on-drop struct since we'll want different logic than
+                // just normally releasing the token in this case.
+                //
+                // On unix this unfortunately means that we lose the specific byte that
+                // was in the pipe (i.e., we just write back the same byte all the time)
+                // but that's not expected to be a problem.
+                std::mem::forget(token.expect("acquire token"));
+
+                let mut tokens = tokens_thread.lock().unwrap();
+                tokens.pending += 1;
```

Comment:
I think that we'll want to remove the concept of pending since it seems like it runs a pretty high risk of collecting too many tokens by accident.

Could we perhaps have a queue of channel pairs? When someone wants a token they allocate a channel and put half the channel into the queue. When a token is acquired a channel is popped off and the token is sent to the client (forgoing the `forget` above as well). That way if we get a token but no thread is still around to use it we'd just drop it on the floor and it'd get released back to the OS.